[GTK][WPE][Skia] Switch UnacceleratedBuffer buffer format to RGBA

nikolaszimmermann opened the pull request:
#### 8a692724af8345c25047dd8ff9620c3eec2673b9
<pre>
[GTK][WPE][Skia] Switch UnacceleratedBuffer buffer format to RGBA
<a href="https://bugs.webkit.org/show_bug.cgi?id=279614">https://bugs.webkit.org/show_bug.cgi?id=279614</a>

Reviewed by Miguel Gomez.

Cairo rendering was performed in BGRA/ARGB (little/big-endian) and
BitmapTexture had to take care of converting to RGBA, when mapping
CPU-rendered buffers to GL textures. Skia supports RGBA directly, saving
the color conversion in TextureMapperShaderProgram.

Testing on desktop/rpi4 shows this change is perf neutral. This is a
preparation to introduce a hybrid rendering mode, where both CPU/GPU
participate in the rendering pipeline, simultaneously.

* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::UnacceleratedBuffer::UnacceleratedBuffer):
(Nicosia::UnacceleratedBuffer::pixelFormat const):
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:
* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::BitmapTexture::BitmapTexture):
(WebCore::BitmapTexture::reset):
(WebCore::BitmapTexture::updateContents):
(WebCore::BitmapTexture::copyFromExternalTexture):
(WebCore::BitmapTexture::colorConvertFlags const):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::TextureMapper::drawNumber):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStoreTile::swapBuffers):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferNativeImage.cpp:
(WebCore::CoordinatedPlatformLayerBufferNativeImage::tryEnsureBuffer):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp:
(WebCore::CoordinatedPlatformLayerBufferVideo::paintToTextureMapper):

Canonical link: <a href="https://commits.webkit.org/283783@main">https://commits.webkit.org/283783@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e2cc0f6c01ff276d603bb299bb4a74a05d9282f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67340 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46719 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19972 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71386 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18477 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69458 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54517 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18268 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53932 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12386 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70407 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42910 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58250 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34466 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39582 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15650 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16835 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61550 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15991 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73087 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11299 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15323 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61458 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11332 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58315 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61495 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14925 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9250 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2854 "Passed tests") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43601 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44787 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43342 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:
```diff
@@ -27,7 +27,7 @@
 
 #pragma once
-
+#include "GraphicsContextGL.h"
```
Can we forward declare the enum instead of including this header? or Use PixelFormat.h instead?

nikolaszimmermann replied:
Ok, PixelFormat.h can be chosen, as well. Eventually we need to extend it for ARGB.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -106,14 +106,26 @@ void BitmapTexture::reset(const IntSize& size, OptionSet<Flags> flags)
 #endif
+}
+
+void BitmapTexture::updateContents(GraphicsContextGL::DataFormat pixelFormat, const void* srcData, const IntRect& targetRect, const IntPoint& sourceOffset, int bytesPerLine)
```
I would put the pixelFormat the last parameters, or at least after the data. Maybe it would be better to force callers to always pass the format and keep a single function.

magomez replied:
Agree. If you don't want to touch all the places where updateContents is used, you can use parameter with the default value BGRA, and only the nicosia buffer will pass RGBA.

nikolaszimmermann replied:
But then we have to change the default value based on, if we're on big/little endian, no?
Because as I understood it's BGRA on little-endian, ARGB on big-endian (for Cairo buffers).

magomez replied:
> But then we have to change the default value based on, if we're on big/little endian, no? Because as I understood it's BGRA on little-endian, ARGB on big-endian (for Cairo buffers).

Or we can separate the format from whether we're big or little endian. Use BGRA or RGBA for the format. What needs to take the endianess into account is the color conversion flag. That's were you need to set BGRAToRGBA on big endian or ARGBToRGBA on little endian if the format is BGRA.

magomez reviewed (changes requested):
We should make the approach a bit more robust taking into account that the BitmapTexture can be reused and updated with content from a different format:
- copyFromExternalTexture always use RGBA, so must always remove the color conversion flag. Before removing it, I would add an assertion checking that either the BitmapTexture doesn't have the colorConversion flag (it's RGBA) or it has it (it's BGRA) but we're updating the complete texture, so we're sure we're not mixing formats.
- the same with updateContents: it must always set the color conversion flag if we're receiving BGRA and always remove it if we're receiving RGBA (you don't need to set the RGBA to RGBA flag, an empty one means no conversion). Like in the previous case, we should assert that either the update fills the whole content or the existent and incoming formats are the same.

magomez commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -104,16 +104,22 @@ void BitmapTexture::reset(const IntSize& size, OptionSet<Flags> flags)
-    // We are updating a texture with format RGBA with content from a buffer that has BGRA format. Instead of turning BGRA
-    // into RGBA and then uploading it, we upload it as is. This causes the texture format to be RGBA but the content to be BGRA,
-    // so we mark the texture to convert the colors when painting the texture.
+    RELEASE_ASSERT(pixelFormat == PixelFormat::RGBA8 || pixelFormat == PixelFormat::BGRA8);
```
The assert idea that I proposed here was to check that we weren't mixing formats. It can happen that a BGRA BitmapTexture gets updated with RGBA content and viceversa, and this is only allowed if we update the whole texture rectangle, cause we're moving all the pixels to the new format. Partial updates can't happen with different formats, and that's we should assert that is not happening.

magomez commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -104,16 +104,22 @@ void BitmapTexture::reset(const IntSize& size, OptionSet<Flags> flags)
+    RELEASE_ASSERT(pixelFormat == PixelFormat::RGBA8 || pixelFormat == PixelFormat::BGRA8);
+
+    if (pixelFormat == PixelFormat::RGBA8)
+        m_colorConvertFlags = { };
```
For example, a single pixel update with RGBA format causes the whole texture to be interpreted as RGBA, which is wrong and will cause visual errors.

magomez commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -309,6 +315,15 @@ void BitmapTexture::copyFromExternalTexture(GLuint sourceTextureID, const IntRec
     RELEASE_ASSERT(sourceOffset.width() + targetRect.width() <= m_size.width());
     RELEASE_ASSERT(sourceOffset.height() + targetRect.height() <= m_size.height());
+#if USE(SKIA)
+    RELEASE_ASSERT(m_colorConvertFlags.isEmpty());
```
Same as the updateContents case, changing the format is not always wrong. It's fine if we update the whole texture. What we need to assert is that we're not changing the format with partial updates. If we don't allow format changes when they cover the whole texture, then a texture that contains one format can never be reused from the pool to hold the other format.

nikolaszimmermann committed:
[GTK][WPE][Skia] Switch UnacceleratedBuffer buffer format to RGBA
https://bugs.webkit.org/show_bug.cgi?id=279614

Reviewed by Miguel Gomez.

Cairo rendering was performed in BGRA/ARGB (little/big-endian) and
BitmapTexture had to take care of converting to RGBA, when mapping
CPU-rendered buffers to GL textures. Skia supports RGBA directly, saving
the color conversion in TextureMapperShaderProgram.

Testing on desktop/rpi4 shows this change is perf neutral. This is a
preparation to introduce a hybrid rendering mode, where both CPU/GPU
participate in the rendering pipeline, simultaneously.

* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::UnacceleratedBuffer::UnacceleratedBuffer):
(Nicosia::UnacceleratedBuffer::pixelFormat const):
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:
* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::BitmapTexture::BitmapTexture):
(WebCore::BitmapTexture::reset):
(WebCore::BitmapTexture::updateContents):
(WebCore::BitmapTexture::copyFromExternalTexture):
(WebCore::BitmapTexture::colorConvertFlags const):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::TextureMapper::drawNumber):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStoreTile::swapBuffers):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferNativeImage.cpp:
(WebCore::CoordinatedPlatformLayerBufferNativeImage::tryEnsureBuffer):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp:
(WebCore::CoordinatedPlatformLayerBufferVideo::paintToTextureMapper):

Canonical link: https://commits.webkit.org/283783@main
